multiselect.go

```diff
@@ -81,20 +87,51 @@ func (m *MultiSelect) OnChange(line []rune, pos int, key rune) (newLine []rune,
 		}
 		// if the user pressed down and there is room to move
 	} else if key == terminal.KeySpace {
-		if old, ok := m.checked[m.Options[m.selectedIndex]]; !ok {
-			// otherwise just invert the current value
-			m.checked[m.Options[m.selectedIndex]] = true
-		} else {
-			// otherwise just invert the current value
-			m.checked[m.Options[m.selectedIndex]] = !old
+		if m.selectedIndex < len(options) {
+			if old, ok := m.checked[options[m.selectedIndex]]; !ok {
+				// otherwise just invert the current value
+				m.checked[options[m.selectedIndex]] = true
+			} else {
+				// otherwise just invert the current value
+				m.checked[options[m.selectedIndex]] = !old
+			}
 		}
 		// only show the help message if we have one to show
 	} else if key == core.HelpInputRune && m.Help != "" {
 		m.showingHelp = true
+	} else if key == terminal.KeyEscape {
+		if m.filter != "" {
```

Comment:
Since we already provide a way to clear the filter, I think it would be best to not overloading the esc behavior. If the user  presses escape, i think we should keep the filter as is, and allow them to navigate the select with j/k. Is that too complicated of an interaction? Knowing that `esc` toggles the input, it was sort of confusing to see the filter disappear when I pressed it.

Reply:
I felt that `Esc` was the most natural way to clear the filter; I wasn't sure how many users would think of hitting `Ctrl+W` or `Ctrl+X`

Reply:
Yea i understand that point, I think since survey already supports j/k for movement, like many other similar packages, I'm trying to find the right balance. While I agree that `ctrl+x` and `ctrl+w` are not super obvious, I want to keep the mental model for the original feature-set as clear as possible. I think having `esc` not clear the filter is the best way to do that. Like I said, as a vim user myself, I was very confused when esc cleared the filter.

Reply:
Yeah - I did wonder with the new magic `ctrl+x` and `ctrl+w` key presses we may need 'keys' help that appears (in addition to any custom help) to help users learn the key presses available?

Reply:
Or to say that another way. If I'm using some CLI tool thats using Survey under the covers and I start filtering a long list; how can I learn the magic `ctrl+x` /  `ctrl+w` key presses to clear the filter? Or learn that there's a vim mode. Figure we should have some way to show some help for the supported keys.

Due to screen real estate it might be we have to link to a little doc via a URL if we can't keep it pretty concise (a couple of lines of the terminal)?

Reply:
Yea this is the exact concern I alluded to in the thread around the VimMode field. We already support Help text with every prompt. Wouldn’t be technically challenging to add navigations docs to those templates in the select/multi select prompts but finding the right copy might be trickier. 

I think whether esc clears the filter or toggles vim mode, we still have this documentation problem. How do you feel about having esc behave like it would in vim, and we address the docs propel either in this PR or another

Reply:
sounds good to me!

Reply:
BTW letting users know they can filter the list might be handy in the help too - it might not be obvious to new users

Reply:
Yea I agree, this need is partly captured in #61, probably worth adding some more info. Maybe even a new ticket that isn't cluttered by all of the other usability improvements that have been made since it's opening.